File: src/main/java/io/github/mjaroslav/bon2gradle/Bon2GradleExtension.java
package io.github.mjaroslav.bon2gradle;

import io.github.mjaroslav.bon2gradle.api.MappingProvider;
import io.github.mjaroslav.bon2gradle.provider.DeobfuscatedDependencyProvider;
import io.github.mjaroslav.bon2gradle.util.LazyNullable;
import io.github.mjaroslav.bon2gradle.util.StringUtils;
import lombok.val;
import org.gradle.api.GradleScriptException;
import org.gradle.api.Project;
import org.gradle.api.file.RegularFileProperty;
import org.gradle.api.provider.Property;
import org.gradle.api.tasks.Internal;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;


public abstract class Bon2GradleExtension {
    private final @NotNull Project project;

    public abstract Property<Boolean> getForceMapping();

    public abstract RegularFileProperty getForcedMappingLocation();

    public abstract Property<String> getForcedMappingRelativeConfPath();

    public abstract Property<String> getMappingProvider();

    public abstract Property<Boolean> getUseParallelDeobfuscation();

    public Bon2GradleExtension(@NotNull Project project) {
        this.project = project;
        getForceMapping().convention(false);
        getMappingProvider().convention(Bon2GradleConstants.DEFAULT_MAPPING_PROVIDER);
        getUseParallelDeobfuscation().convention(false);
    }

    private final LazyNullable<MappingProvider> mappingProviderObject = new LazyNullable<>(() -> {
        val className = getMappingProvider().get();
        if (StringUtils.isEmpty(className)) {
            Bon2GradlePlugin.getLogger().info("MappingProvider not provided");
            return null;
        }
        try {
            return (MappingProvider) Class.forName(className).getConstructor().newInstance();
        } catch (Exception e) {
            throw new GradleScriptException("Error while mapping provider instantiation", e);
        }
    });

    @Internal
    public @Nullable MappingProvider getMappingProviderObject() {
        return mappingProviderObject.get();
    }

    @Contract("_ -> new")
    public DeobfuscatedDependencyProvider deobf(@NotNull Object identifier) {
        return new DeobfuscatedDependencyProvider(project, identifier);
    }
}

File: src/main/kotlin/Bon2GradleExtension.kt
package io.github.mjaroslav.bon2gradle

import io.github.mjaroslav.bon2gradle.api.MappingProvider
import io.github.mjaroslav.bon2gradle.provider.DeobfuscatedDependencyProvider
import org.gradle.api.GradleScriptException
import org.gradle.api.Project
import org.gradle.api.file.RegularFileProperty
import org.gradle.api.provider.Property
import javax.inject.Inject

abstract class Bon2GradleExtension @Inject constructor(private val project: Project) {
    val forceMapping: Property<Boolean> = project.objects.property(Boolean::class.java).convention(false)
    val forcedMappingLocation: RegularFileProperty = project.objects.fileProperty()
    val forcedMappingRelativeConfPath: Property<String> = project.objects.property(String::class.java).convention("")
    val mappingProvider: Property<String> =
        project.objects.property(String::class.java).convention(DEFAULT_MAPPING_PROVIDER)
    val useParallelDeobfuscation: Property<Boolean> = project.objects.property(Boolean::class.java).convention(false)

    val mappingProviderObject: MappingProvider? by lazy {
        val result: MappingProvider?
        val className = mappingProvider.get()
        if (className.isBlank()) project.logger.info("MappingProvider not provided")
        try {
            result = Class.forName(className).getConstructor().newInstance() as MappingProvider
        } catch (e: Exception) {
            throw GradleScriptException("Error while mapping provider instantiation", e)
        }
        result
    }

    fun deobf(identifier: Any): DeobfuscatedDependencyProvider =
        DeobfuscatedDependencyProvider(project, identifier)
}
